src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -66,6 +66,11 @@ public void Add(string hintName, SourceText source)
                 throw new ArgumentException(CodeAnalysisResources.HintNameUniquePerGenerator, nameof(hintName));
             }
 
+            if (source.Encoding is null)
+            {
+                throw new ArgumentException(CodeAnalysisResources.SourceTextRequiresEncoding, nameof(source));
+            }
```

Comment:
❔ Why not make it either inherit the encoding for the rest of the project files (per .editorconfig), or automatically assign `Encoding.UTF8`? The easiest path might be to add an internal `WithEncoding` method, which delegates to the original `SourceText` for everything but the `Encoding` property.

Reply:
Chatted with @tmat and @jaredpar about this. Basically if you want automatic encoding selection, use the string overload. If you're passing a `SourceText` then we're saying you have to tell us what the encoding is (we don't know if it came from a string) and so we want to fail early. 

We can always open it up to make it _less_ strict later on, and start taking a default encoding, we can't make it more strict.

Reply:
works for me